packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -349,24 +330,115 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
+                if (!entry) {
+                  return currentRelationData;
+                }
+
+                const relatedEntry = await strapi.db
+                  .query('plugin::upload.file')
+                  .findOne({ where: { id: entry.id } });
+
+                if (relatedEntry) {
+                  currentRelationData.results.push(relatedEntry);
+                } else {
+                  // The related content has been deleted
+                  currentRelationData.meta.missingCount += 1;
+                }
+
+                return currentRelationData;
+              },
+              Promise.resolve({
+                results: [] as any[],
+                meta: { missingCount: 0 },
+              })
+            )
+        );
+      };
+
       const populateEntryRelations = async (
         entry: HistoryVersionQueryResult
       ): Promise<CreateHistoryVersion['data']> => {
         const entryWithRelations = await Object.entries(entry.schema).reduce(
           async (currentDataWithRelations, [attributeKey, attributeSchema]) => {
+            const attributeValue = entry.data[attributeKey];
+            const attributeValues = Array.isArray(attributeValue)
+              ? attributeValue
+              : [attributeValue];
+
+            if (attributeSchema.type === 'media') {
+              return {
+                ...(await currentDataWithRelations),
+                [attributeKey]: await buildMediaResponse(attributeValues),
+              };
+            }
+
             // TODO: handle relations that are inside components
-            if (['relation', 'media'].includes(attributeSchema.type)) {
-              const attributeValue = entry.data[attributeKey];
-              const relationResponse = await buildRelationReponse(
-                (Array.isArray(attributeValue)
-                  ? attributeValue
-                  : [attributeValue]) as EntryToPopulate[],
-                attributeSchema
-              );
+            if (
+              attributeSchema.type === 'relation' &&
+              attributeSchema.relation !== 'morphToOne' &&
+              attributeSchema.relation !== 'morphToMany'
+            ) {
+              /**
+               * Don't build the relations response object for relations to admin users,
+               * because pickAllowedAdminUserFields sanitizati
+               */
+              if (attributeSchema.target === 'admin::user') {
+                const sanitizedAdminUser = await Promise.all(
+                  attributeValues.map(async (userToPopulate) => {
+                    if (userToPopulate == null) {
+                      return null;
+                    }
+
+                    const user = await strapi
+                      .query('admin::user')
+                      .findOne({ where: { id: userToPopulate.id } });
+
+                    const permissionChecker = getContentManagerService('permission-checker').create(
```

Comment:
You should move this outside of the loop

Reply:
Good thinking!